Support command line options in EDITOR/GIT_EDITOR

nirs opened the pull request:
On Windows users may like to configure command line options for the
editor[1]. Support this configuration.

Fixes #162.

[1] https://git-scm.com/book/en/v2/Getting-Started-First-Time-Git-Setup

nirs commented:
@zware, @larryhastings please review.

nirs commented:
Here an example with command with spaces, and command line options:
```
ln -s /usr/bin/vim 'my_editor'
$ EDITOR="$PWD/'my editor' +25" ../../core-workflow/blurb/blurb.py
```
It opens vim and jump to line 25.

Not tested on Windows.

larryhastings commented:
Yes, but you use quotes around the spaces.  If the path isn't quoted, it won't run the editor properly.  I don't know whether or not people put unquoted paths with spaces in environment variables, but I think it's possible to handle it--please see my previous pseudocode.

nirs commented:
I have seen your example in #156, but it will also not handle path with spaces *and* command line arguments. If you don't quote the path, we get:

```python
>>> p = "C:/Program Files (x86)/Notepad++/notepad++.exe -multiInst -nosession"
>>> shlex.split(p, posix=False)
['C:/Program', 'Files', '(x86)/Notepad++/notepad++.exe', '-multiInst', '-nosession']
```
We can use this hack for windows:
```python
>>> p[:p.index('.exe') + 4]
'C:/Program Files (x86)/Notepad++/notepad++.exe'
>>> shlex.split(p[p.index('.exe') + 4:])
['-multiInst', '-nosession']
```

larryhastings commented:
It's impossible** to handle a string that contains *both* a path with unquoted spaces *and* additional command-line arguments.  My point was, I think we need to handle these two scenarios:

1. $EDITOR is a unquoted path to an editor, containing spaces
2. $EDITOR is a quoted path to an editor, followed by arguments

My proposed approach would handle both those cases; IIUC your current approach only handles the latter.

I think it's reasonable to handle both these scenarios. After all, there are plenty of examples of environment variables that can contain unquoted paths containing spaces (e.g. $PATH, or %APPDATA% on Windows).  And there are also examples specifically of $EDITOR containing command-line arguments, as you have demonstrated.

Therefore, I don't want to merge your PR with its current approach.  If you'd like to handle this more complicated logic, go ahead and try; otherwise we should abandon this PR and I'll do the more complicated thing.

** well, not *impossible* to handle such a string. but it would be ridiculous to try.  let's not go there.

methane commented:
Mercurial seems doing like tihs:

```
os.system('%s "%s"' % (cmd, path))
```

See https://www.mercurial-scm.org/repo/hg/file/tip/mercurial/ui.py#l1379

nirs commented:
@larryhastings next patch implements what you suggested, supporting both ```"unquoted path"``` and ```"'quoted apth' --with --options"```.

@methane Mercurial solution is elegant, but it will not work with unquoted path.

nirs commented:
@larryhastings, can you review the latest version again?

nirs commented on blurb/blurb.py:
```diff
@@ -864,7 +882,7 @@ def find_editor():
             found_path = shutil.which(fallback)
         if found_path and os.path.exists(found_path):
-            return found_path
+            return [found_path]
```
Yes, this looks like a bug fix, I can split it to another PR if you like.

nirs commented on blurb/blurb.py:
```diff
@@ -874,11 +892,12 @@ def add():
     handle, tmp_path = tempfile.mkstemp(".rst")
     os.close(handle)
     atexit.register(lambda : os.unlink(tmp_path))
+    editor_command.append(tmp_path)
```
editor_command is used in line 837. I don't see the append you mention. I'll need to rebase and test this again, lot of time has passed since I sent this :-)

larryhastings commented:
This PR seems to be languishing.  That's fine, but I *would* like the `return [found_path]` bug fix on line 885 to get merged.  As you suggest it should arguably be a separate PR anyway.

If you have the time I'd be happy for you to make a one-line PR to get that fix in, so you can get credit for it.  Or I can just do it if you don't care about getting credit.
